### PR TITLE
Support no_proxy via URI::Generic#find_proxy

### PR DESCRIPTION
**Pull request by authorNari**
Default proxy value is set via `URI::Generic#find_proxy` if it's defined.
Maintenance cost is lower using upstream implementation than a implementation is made by myself.

### CODE REVIEW

File: lib/faraday/connection.rb [hunk 1]
```diff
@@ -464,5 +442,16 @@ def set_authorization_header(header_type, *args)
...
+
+    def find_default_proxy
+      warn 'no_proxy is unsupported' if ENV['no_proxy'] || ENV['NO_PROXY']
+      uri = ENV['http_proxy']
+      if uri && !uri.empty?
+        uri = 'http://' + uri if uri !~ /^http/i
+        return uri
+      else
+        return nil
+      end
```
Comment (olleolleolle) [inline lib/faraday/connection.rb]: Code note: drop the `return` keyword on lines 451 and 453 (the `if`'s the last expression of the method).

**Comment by iMacTia** [comment]
Hi @authorNari,

I can see many tests have been deleted in your PR.
Can I ask you why?

**Comment by authorNari** [comment]
Hi @iMacTia 

511a5b4 is reverted on this pull request so these tests are deleted.
These tests are depends on `proxy_allowed?` method.

File: test/connection_test.rb [hunk 2]
```diff
@@ -326,126 +319,71 @@ def test_proxy_accepts_hash
...
-      conn.proxy :uri => :bad_uri, :user => 'rick'
-    end
-  end
-
-  def test_proxy_allowed_when_url_in_no_proxy_list
-    with_env 'http_proxy' => 'http://proxy.com', 'no_proxy' => 'example.com' do
-      conn = Faraday::Connection.new
-      assert_equal conn.proxy_allowed?('http://example.com'), false
-    end
-  end
```
Comment (iMacTia) [inline test/connection_test.rb]: Please reintroduce all the deleted test as they're covering some important edge-cases.
This can be done without using the `proxy_allowed?` method:

- If the test asserts `assert_equal conn.proxy_allowed?('http://example.com'), false`, then you can change it into `assert_nil conn.proxy`
- If the test asserts `assert_equal conn.proxy_allowed?('http://prefixedexample.com'), true` than you can change it into `assert_equal 'proxy.com', conn.proxy.host`

File: lib/faraday/connection.rb [hunk 3]
```diff
@@ -79,15 +79,22 @@ def initialize(url = nil, options = nil)
...
 
-      @proxy = @no_proxy = nil
+      @proxy = nil
       proxy(options.fetch(:proxy) {
-        uri = ENV['http_proxy']
-        if uri && !uri.empty?
-          uri = 'http://' + uri unless uri.downcase.start_with?('http')
-          uri
+        uri = nil
+        if URI.parse("").respond_to?(:find_proxy)
```
Comment (iMacTia) [inline lib/faraday/connection.rb]: According to the documentation, `find_proxy` is supported in Ruby 1.9.3 (https://ruby-doc.org/stdlib-1.9.3/libdoc/open-uri/rdoc/URI/Generic.html). Since we dropped the support for older ruby version, I think this method should always be available? Do you get any failing test if you remove this check?

File: lib/faraday/connection.rb (hunk 1 above)
Comment (iMacTia) [inline lib/faraday/connection.rb]: I agree. Moreover, the `else` branch is completely redundant. so the `if` can be simplified into:

```ruby
if uri && !uri.empty?
  uri = 'http://' + uri if uri !~ /^http/i
  uri
end
```

File: test/connection_test.rb (hunk 2 above)
Comment (authorNari) [inline test/connection_test.rb]: That make sense. done bbd89c1

File: lib/faraday/connection.rb (hunk 3 above)
Comment (authorNari) [inline lib/faraday/connection.rb]: >  Since we dropped the support for older ruby version, I think this method should always be available?

`open-uri` lib has `find_proxy` in >= Ruby 1.8.7. This method move to `uri/generic` lib in Ruby 2.0.0.
I think any library shouldn't require `open-uri` because this library overwrite Kernel#.open.
So this pull request doesn't support no_proxy in Ruby 1.9.3. 😢

File: lib/faraday/connection.rb (hunk 1 above)
Comment (authorNari) [inline lib/faraday/connection.rb]: I think If the return value of a method is meaningful, `return` should be explicitly written. It's easy to read :)

But I think that the consistency of this project should be respected.
I deleted `return` dd13327

File: lib/faraday/connection.rb (hunk 3 above)
Comment (iMacTia) [inline lib/faraday/connection.rb]: Thanks for the detailed response @authorNari that makes totally sense.
I have nothing against this as the plan is to move away from Ruby1.9.3 as soon as possible, so this can be a good incentive!

**Comment by iMacTia** [comment]
Thank you @authorNari, outstanding work!
I'm grateful we could delegate to Ruby the management of the proxy/no_proxy ENV variables check.
Apologies if it took me some time to review and thanks again for the valuable contribution.

File: lib/faraday/connection.rb [hunk 4]
```diff
@@ -464,5 +442,14 @@ def set_authorization_header(header_type, *args)
         header(*args)
       headers[Faraday::Request::Authorization::KEY] = header
     end
+
+    def find_default_proxy
+      warn 'no_proxy is unsupported' if ENV['no_proxy'] || ENV['NO_PROXY']
```
Comment (somethingnew2-0) [inline lib/faraday/connection.rb]: So no_proxy is unsupported when the url is not given upfront in the initializer, but later on in the request?

Comment (iMacTia) [inline lib/faraday/connection.rb]: You mean when you initialise the connection simply as `conn = Faraday.new` ?
That is actually true, I missed this detail as it's an unusual way to use it.
This could be easily fixed moving the #find_proxy call on the `run_request` or `build_request` methods.

Comment (somethingnew2-0) [inline lib/faraday/connection.rb]: Yes, I agree it is a weird way to use it, but it is allowed.  The find_proxy method should be moved to when the request is created in `build_request`.

Comment (iMacTia) [inline lib/faraday/connection.rb]: Totally agree with you, I'll try to squeeze this small fix in release 0.12.0 👍
